forte/data/extractor/utils.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains utility functions for extractors.
+"""
+
+from typing import List, Tuple
+from ft.onto.base_ontology import Annotation, Token
+
+
+def bio_tagging(pack, instance_based_on: List[Annotation],
+                instance_entry: List[Annotation]) \
+                -> List[Tuple[Annotation, str]]:
+    """This utility function use BIO tagging method to convert tags
+    of "instance_entry" into the same length as "instance_based_on". Both
+    element from "instance_entry" and "instance_based_on" should Annotation
+    type. This function uses their position information to
+    determine B, I, O tagging for the entry on each based_on element.
+
+    Args:
+        instance_based_on (List[Annotation]): The Annotations that
+            tags need to align to.
+
+        instance_entry (List[Annotation]): Annotations where tags come from.
+
+    Returns:
+        A list of the type List[Tuple[Annotation, str]]. For example,
+        [[None, "O"], [tag1, "B"], [tag1, "I"], [None, "O"],
+         [None, "O"], [tag2, "B"], [None, "O"]]
```

Comment:
Why don't we pass in the attribute key and get it here as well? So we return ('LOC', 'B'). or simply `B-LOC`

Reply:
1. The attribute might not be str, it is more flexible to put them into a tuple.
2. "B-LOC" assumes the delimiter “-” does not exist in the attribute, otherwise when we convert "B-LOC" back to "B" and "LOC", there will be problem and it is inconvenient to do this in the `add_to_pack` function.

Reply:
I misunderstood the question. I think returning attribute "LOC" is also ok here. But I return the Annotation here.

Maybe it is more flexible to return Annotation first and then get the attribute latter. For example, if there is some different ways get the attribute just like that in the AttributeExtractor.

I have no special reason for this.

Reply:
I think returning `LOC` is more intuitive. A special way to get the attribute is a good point. One a bit complex way to do this is passing in a Union of function and str: `attribute: Union[callable[[Annotation], str], str]`, and do this:

if isinstance(attribute, callable):
   tag = attribute(unit)
else:
   tag = getattr(unit, str)

If you want to have better type-check, you can try using this:
https://github.com/asyml/forte/blob/master/forte/utils/utils.py#L148

Reply:
Got it.

Reply:
```
  data = []
  for pair in instance_tagged:
      if pair[0] is None:
          new_pair = (None, pair[1])
      else:
          new_pair = (getattr(pair[0], self.config.attribute), pair[1])
      if self.vocab:
          data.append(self.element2repr(new_pair))
      else:
          data.append(new_pair)
```